V16: Documents Fetch and http-client in the Backoffice (#7083)
umbraco/UmbracoDocs: 16/umbraco-cms/customizing/foundation/fetching-data/http-client.md

---
description: Learn more about working with the Umbraco HTTP Client.
---

# HTTP Client

> The Umbraco Backoffice provides a built-in HTTP client that you can use to make network requests. This client is colloquially known as the Umbraco HTTP Client. It is generated using **@hey-api/openapi-ts** around the OpenAPI specification and is available through the `@umbraco-cms/backoffice/http-client` package.

Reviewer: ```suggestion
The Umbraco Backoffice includes a built-in HTTP client commonly referred to as the Umbraco HTTP Client for making network requests. It is generated using `@hey-api/openapi-ts` around the OpenAPI specification and is available through the `@umbraco-cms/backoffice/http-client` package.
```

**Example:**

```javascript
import { umbHttpClient } from '@umbraco-cms/backoffice/http-client';

const { data } = await umbHttpClient.get({
	url: '/umbraco/management/api/v1/server/status'
});

if (data) {
	console.log('Server status:', data);
}
```

The above example shows how to use the Umbraco HTTP client to make a GET request to the Management API. The `umbHttpClient` object provides methods for making requests, including `get`, `post`, `put`, and `delete`. Each method takes an options object that specifies the URL, headers, and body of the request.

The Umbraco HTTP client automatically handles authentication and error handling, so you don't have to worry about those details. It also provides a convenient way to parse the response data as JSON.

## Using the Umbraco HTTP Client

> The Umbraco HTTP client is a wrapper around the Fetch API that provides a more convenient way to make network requests. It handles things like request and response parsing, error handling, and retries. The Umbraco HTTP client is available through the `@umbraco-cms/backoffice/http-client` package, which is included in the Umbraco Backoffice. You can use it to make requests to any endpoint in the Management API or to any other API.

Reviewer: ```suggestion
The Umbraco HTTP client is a wrapper around the Fetch API that provides a more convenient way to make network requests. It handles request and response parsing, error handling, and retries. The Umbraco HTTP client is available through the `@umbraco-cms/backoffice/http-client` package, which is included in the Umbraco Backoffice. You can use it to make requests to any endpoint in the Management API or to any other API.
```

The recommended approach to use the Umbraco HTTP Client is to use the `tryExecute` function. This function will handle any errors that occur during the request and will automatically refresh the token if it is expired. If the session is expired, the function will also make sure the user logs in again.

You can read more about the `tryExecute` function in this article:

> {% content-ref url="../try-execute.md" %}
> [try-execute.md](../try-execute.md)
> {% endcontent-ref %}

Reviewer 2: ```suggestion
{% content-ref url="try-execute.md" %}
[Executing Requests](try-execute.md)
{% endcontent-ref %}
```
Corrected link path (it's in the same directory).

## Generating a Custom Client

> You can also generate your own client using the **@hey-api/openapi-ts** library. This library allows you to generate a TypeScript client from an OpenAPI specification. The generated client will handle authentication and error handling for you, so you don't have to worry about those details.

Reviewer: ```suggestion
You can also generate your own client using the `@hey-api/openapi-ts` library. This library allows you to generate a TypeScript client from an OpenAPI specification. The generated client will handle authentication and error handling for you, so you don't have to worry about those details.
```

Read more about generating your own client here:

{% content-ref url="custom-generated-client.md" %}

> [custom-generated-client.md](custom-generated-client.md)

Reviewer 2: ```suggestion
[Custom Generated Client](custom-generated-client.md)
```
Friendlier link label?

{% endcontent-ref %}

## Further reading

> - `[@hey-api/openapi-ts](https://heyapi.dev/openapi-ts/get-started)`

Reviewer: ```suggestion
- [@hey-api/openapi-ts](https://heyapi.dev/openapi-ts/get-started)
```

> - [@hey-api/openapi-ts](https://heyapi.dev/openapi-ts/get-started)

Reviewer: ```suggestion
- `[@hey-api/openapi-ts](https://heyapi.dev/openapi-ts/get-started)`
```

> - `[@umbraco-cms/backoffice/http-client](https://apidocs.umbraco.com/v16/ui-api/modules/packages_core_http-client.html)`

Reviewer: ```suggestion
- [@umbraco-cms/backoffice/http-client](https://apidocs.umbraco.com/v16/ui-api/modules/packages_core_http-client.html)
```

> - [@umbraco-cms/backoffice/http-client](https://apidocs.umbraco.com/v16/ui-api/modules/packages_core_http-client.html)

Reviewer: ```suggestion
- `[@umbraco-cms/backoffice/http-client](https://apidocs.umbraco.com/v16/ui-api/modules/packages_core_http-client.html)`
```
